[Enhancement] Support multiple compression formats (GZIP/SNAPPY/ZSTD/LZ4) for CSV file exports

tracymacding opened the pull request:
Add GZIP/SNAPPY/ZSTD/LZ4 compression support to CSV exports via INSERT INTO FILES. Users can now specify 'compression'='gzip' to compress CSV output files, which automatically generates .csv.gz files.

**Key changes:**

1. Add CompressedAsyncOutputStreamFile class to handle compression
2. Update CSVFileWriter to support compressed output streams
3. Automatically append .gz extension for compressed CSV files
4. Add comprehensive unit and integration tests (97.8% coverage)
5. Use BlockCompressionCodec for efficient compression

**Usage example**

```
INSERT INTO FILES(
    'path' = 'file:///tmp/export/',
    'format' = 'csv',
    'compression' = 'gzip'
)
SELECT * FROM my_table;
```

## Why I'm doing:

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.1
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables compressed CSV exports and auto file suffixing, with a new stream-based compression layer and updated writer factory integration.
> 
> - Introduces `csv::CompressedOutputStream` (decorator over `OutputStream`) supporting GZIP/SNAPPY/ZSTD/LZ4; incremental chunking for GZIP, buffer-then-compress for others
> - Updates `CSVFileWriterFactory::create` to wrap the async stream with compression based on `TCompressionType`, removing compression logic from `CSVFileWriter`
> - Appends compression extensions to CSV file names in `FileChunkSinkProvider` (`.gz`, `.snappy`, `.zst`, `.lz4`); Parquet/ORC unchanged
> - Adds new `output_stream_file.{h,cpp}` and wires into build; minor refactors in CSV header-writing behavior
> - Extensive tests: new compression utilities, `output_stream_file_test`, and expanded `csv_file_writer_test` covering compression, headers, options, and factory
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c28e129274a205b37b0ed77ff172f6f522443171. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

kevincai commented on be/src/formats/csv/output_stream_file.cpp:
```diff
@@ -0,0 +1,102 @@
+    memcpy(_uncompressed_buffer.data() + old_size, data, size);
+
+    // For compression types that support incremental compression (GZIP),
+    // flush when buffer reaches chunk size to avoid excessive memory usage.
```
it doesn't make sense to hold all the data in memory and do the final compression at the end, e.g. hold a data with 1GB in memory and compress to lz4/snappy/zstd. must be incremental compression with compress ctx.

kevincai committed:
[UT] fix proc profile sql test (#67758)

Signed-off-by: Kevin Cai <kevin.cai@celerdata.com>

tracymacding committed:
[Enhancement] Support compression for CSV file exports

This PR adds compression support for CSV file exports with the following features:

- Add CompressedOutputStream class for handling compressed CSV output
- Support incremental compression via frame concatenation for: GZIP, LZ4_FRAME, ZSTD
- SNAPPY and LZ4 (raw) are not supported as they don't support frame concatenation
- Add appropriate file extensions for compressed files: .gz, .zst, .lz4
- Add comprehensive unit tests for compression functionality
- Fix file cleanup on compression failure using DeferOp
- Fix size() semantics to return accurate compressed bytes estimate

Signed-off-by: tracymacding <tracymacding@gmail.com>

tracymacding committed:
[Enhancement] Improve test coverage for CSV compression

- Fix TestSizeMethod to call size() before finalize() to cover pending
  estimate calculation path
- Remove unused _supports_incremental_compression() method (dead code)

Signed-off-by: tracymacding <tracymacding@gmail.com>

tracymacding committed:
[BugFix] Fix size() to account for base class buffer

The CompressedOutputStream::size() method now correctly accounts for
pending data in both:
1. The base class OutputStream buffer (data written but not yet synced)
2. The _uncompressed_buffer (data synced but not yet compressed)

This fixes a bug where size() returned 0 after writing small amounts
of data because the data was still in the base class buffer.

Signed-off-by: tracymacding <tracymacding@gmail.com>

StarRocks-Reviewer commented:
@cursor review

kevincai commented on be/src/formats/csv/output_stream_file.cpp:
```diff
@@ -0,0 +1,140 @@
+        // ZSTD supports concatenation of multiple frames.
+        return true;
+    default:
+        // SNAPPY, LZ4 (raw block), and others don't support frame concatenation.
```
https://github.com/lz4/lz4/blob/dev/examples/fileCompress.c#L34 

Example on how to compress a large file with chunks.

kevincai commented on be/src/formats/csv/output_stream_file.h:
```diff
@@ -63,4 +66,57 @@ class AsyncOutputStreamFile final : public OutputStream {
+//     standard decompression tools can handle.
+//   - SNAPPY, LZ4 (raw): Do NOT support incremental compression. Creating a
+//     CompressedOutputStream with these types will return an error.
+class CompressedOutputStream final : public OutputStream {
```
this class shall be moved out of csv namespace, it is not csv specific function.
